### PR TITLE
Added flag to explicitly enable zone-awareness replication and added store-gateway support

### PR DESCRIPTION
**Pull request by pracucci**
**What this PR does**:
In this PR:
- As agreed in the discussion we had in the PR #3090, I've added a flag to explicitly enable zone-awareness replication
- Added zone-awareness replication support to store-gateway
- Updated doc

**Which issue(s) this PR fixes**:
N/A

**Checklist**
- [x] Tests updated
- [x] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: pkg/ring/ring.go [hunk 1]
```diff
@@ -402,72 +407,6 @@ func (r *Ring) Collect(ch chan<- prometheus.Metric) {
 	)
 }
 
-// Subring returns a ring of n ingesters from the given ring. If the subring can't be built
-// (ie. because there are not enough instances) then it returns the full ring.
-func (r *Ring) Subring(key uint32, n int) ReadRing {
```
Comment (pracucci) [inline pkg/ring/ring.go:407]: Forgot to remove it in #3090, so removing it now.

File: docs/blocks-storage/store-gateway.template [hunk 2]
```diff
@@ -56,6 +56,16 @@ To protect from this, when an healthy store-gateway instance finds another insta
...
 
+### Zone-awareness
+
+The store-gateway replication optionally supports [zone-awareness](../guides/zone-replication.md). When zone-aware replication is enabled and the blocks replication factor is > 1, each block is guaranteed to be replicated across store-gateway instances running in different availability zones.
+
+**To enable** the zone-aware replication for the store-gateways you should:
+
+1. Configure the availability zone for each store-gateway via the `-store-gateway.sharding-ring.instance-availability-zone` CLI flag (or its respective YAML config option)
+2. Enable blocks zone-aware replication via the `-store-gateway.sharding-ring.zone-awareness-enabled` CLI flag (or its respective YAML config option). Please be aware this configuration option should be set to store-gateways, queriers and rulers.
+3. Rollout store-gateways and queriers to apply the new configuration
```
Comment (pstibrany) [inline docs/blocks-storage/store-gateway.template:67]:
```suggestion
3. Rollout store-gateways, queriers and rulers to apply the new configuration
```

File: docs/guides/zone-replication.md [hunk 3]
```diff
@@ -5,26 +5,39 @@ weight: 5
 slug: zone-aware-replication
 ---
 
-In a default configuration, time-series written to ingesters are replicated based on the container/pod name of the ingester instances. It is completely possible that all the replicas for the given time-series are held with in the same availability zone, even if the cortex infrastructure spans multiple zones within the region. Storing multiple replicas for a given time-series poses a risk for data loss if there is an outage affecting various nodes within a zone or a total outage.
+Cortex supports data replication for different services. By default, data is transparently replicated across the whole pool of service instances, regardless these instances are all running within the same availability zone (or data center, or rack) or in different ones.
```
Comment (pstibrany) [inline docs/guides/zone-replication.md:8]:
```suggestion
Cortex supports data replication for different services. By default, data is transparently replicated across the whole pool of service instances, regardless of whether these instances are all running within the same availability zone (or data center, or rack) or in different ones.
```

File: pkg/ring/ring_test.go [hunk 4]
```diff
@@ -124,16 +123,129 @@ func TestAddIngesterReplacesExistingTokens(t *testing.T) {
...
+			replicationFactor:    3,
+			zoneAwarenessEnabled: true,
+			expectedIngesters:    3,
+		},
+		"should fail if there are ingesters in 1 zone only on RF = 3": {
+			numIngesters:         16,
+			numZones:             1,
+			replicationFactor:    3,
+			zoneAwarenessEnabled: true,
+			expectedErr:          "at least 2 live replicas required, could only find 1",
```
Comment (pstibrany) [inline pkg/ring/ring_test.go:150]: Would it be easy for error message to mention in this case that replicas must be from different zones?

File: pkg/storegateway/gateway_ring.go [hunk 5]
```diff
@@ -67,13 +69,15 @@ func (cfg *RingConfig) RegisterFlags(f *flag.FlagSet) {
...
+	f.BoolVar(&cfg.ZoneAwarenessEnabled, "store-gateway.sharding-ring.zone-awareness-enabled", false, "True to enable zone-awareness and replicate blocks across different availability zones.")
 
 	// Instance flags
 	cfg.InstanceInterfaceNames = []string{"eth0", "en0"}
 	f.Var((*flagext.StringSlice)(&cfg.InstanceInterfaceNames), "store-gateway.sharding-ring.instance-interface", "Name of network interface to read address from.")
 	f.StringVar(&cfg.InstanceAddr, "store-gateway.sharding-ring.instance-addr", "", "IP address to advertise in the ring.")
 	f.IntVar(&cfg.InstancePort, "store-gateway.sharding-ring.instance-port", 0, "Port to advertise in the ring (defaults to server.grpc-listen-port).")
 	f.StringVar(&cfg.InstanceID, "store-gateway.sharding-ring.instance-id", hostname, "Instance ID to register in the ring.")
+	f.StringVar(&cfg.InstanceZone, "store-gateway.sharding-ring.instance-availability-zone", "", "The availability zone of the host, this instance is running on. Required if zone-awareness is enabled.")
```
Comment (pstibrany) [inline pkg/storegateway/gateway_ring.go:80]: "The availability zone where this instance is running. Required if zone-awareness is enabled."

Would this still make sense?

**Review by pstibrany — APPROVED** [review]
Great job!

File: pkg/ring/ring_test.go (hunk 4 above)
Comment (pracucci) [reply pkg/ring/ring_test.go]: Sure, makes sense. Done.

**Comment by pracucci** [comment]
@pstibrany Thanks for your review. Could you check it again, please?

**Review by pstibrany — APPROVED** [review]
LGTM, great job! I especially like ring tests unification :)

**Pull request merged**